src/test/java/com/google/sps/servlets/AddSchoolDataTest.java

```diff
@@ -0,0 +1,449 @@
+// // Copyright 2019 Google LLC
+// //
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets.second;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.when;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import java.util.List;
+import javax.servlet.http.HttpServletRequest;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+/** */
+public final class AddSchoolDataTest {
+
+  private static final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+  private AddSchoolData newSchoolObject;
+
+  @BeforeEach
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+    newSchoolObject = new AddSchoolData();
+  }
+
+  @AfterEach
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Mock HttpServletRequest request;
+
+  @Mock HttpServletRequest requestB;
+
+  @Test
+  public void AddingNewSchool_CreatesNewEntityForAll() {
```

Comment:
Something we forgot to talk about that would help me a lot when reviewing this.

Typically for each test you'll have three sections: Setup, Act, and Assert.

It may even be useful to call them out explicitly via comment, though typically we just put a blank line between.

What this means for your code here:

Setup is creating the request mock, and any variables needed to check against later (like expected entities/lengths/etc)

Act is actually doing the action, in this case just calling the function on your AddSchoolInfo class.

Assert is anything with assertEquals/etc, where you're verifying the behavior.

You have all the right stuff here, it'd just be easier to read if it was moved into that order.

Reply:
Also sometimes you might have a pre-assert section where you make sure the DB is empty/etc. Though in this case since you're mocking it out each time I don't think that's necessary.